pennylane/qchem/vibrational/bosonic.py

```diff
@@ -0,0 +1,639 @@
+        will fail to multiply with a BoseWord"""
+
+        return self.__mul__(other)
+
+    def __pow__(self, value):
+        r"""Exponentiate a Bose word to an integer power.
+
+        >>> w = BoseWord({(0, 0) : '+', (1, 1) : '-'})
+        >>> print(w**3)
+        b⁺(0) b(1) b⁺(0) b(1) b⁺(0) b(1)
+        """
+
+        if value < 0 or not isinstance(value, int):
+            raise ValueError("The exponent must be a positive integer.")
+
+        operator = BoseWord({})
+
+        for _ in range(value):
+            operator *= self
+
+        return operator
+
+    def normal_order(self):
+        r"""Convert a BoseWord to its normal-ordered form.
+
+        >>> bw = BoseWord({(0, 0): "-", (1, 0): "-", (2, 0): "+", (3, 0): "+"})
+        >>> print(bw.normal_order())
+        4.0 * b⁺(0) b(0)
+        + 2.0 * I
+        + 1.0 * b⁺(0) b⁺(0) b(0) b(0)
+
+        .. details::
+            :title: Developer Notes
+
+            This recursive function uses a two pointer approach. The left pointer is positioned at
+            the leftmost annihilation `"-"` term and the right pointer is incremented until it
+            reaches the leftmost creation `"+"` term. The function then does consecutive adjacent
+            swaps on the creation term and the term to its left until the creation term is at the
+            left pointer. Any commutation terms picked up after is then normal ordered, hence the
+            recursion. Finally, the left pointer increments.
+        """
+        bw_terms = sorted(self)
+        len_op = len(bw_terms)
+        bw_comm = BoseSentence({BoseWord({}): 0.0}, is_hardcore=self.is_hardcore)
+
+        if len_op == 0:
+            return 1 * BoseWord({})
+
+        bw = self
+
+        l = 0
+        for r in range(len_op):
+            if self[bw_terms[r]] == "+":
+                if l == r:
+                    l += 1
+                    continue
+
+                bs = bw.shift_operator(r, l)
+                bs_as_list = sorted(list(bs.items()), key=lambda x: len(x[0].keys()), reverse=True)
+                bw = bs_as_list[0][0]
+                # Sort by ascending index order
+                if l > 0:
+                    bw_as_list = sorted(list(bw.keys()))
+                    if bw_as_list[l - 1][1] > bw_as_list[l][1]:
+                        temp_bs = bw.shift_operator(l - 1, l)
+                        bw = list(temp_bs.items())[0][0]
+
+                for i in range(1, len(bs_as_list)):
+                    bw_comm += bs_as_list[i][0] * bs_as_list[i][1]
+
+                l += 1
+
+        ordered_op = bw + bw_comm.normal_order()
+        ordered_op.simplify(tol=1e-8)
+
+        if self.is_hardcore:
+            for bw, _ in ordered_op.items():
+                bw_array = list(bw.keys())
+                indice_array = [x[1] for x in bw_array]
+                if len(indice_array) != len(set(indice_array)):
+                    ordered_op[bw] = 0
```

Comment:
Please make sure that the codecov issue is properly resolved.

Reply:
I don't see this codecov issue. The PR has all checks passing, so I am not really sure what you're seeing...

Reply:
Oh ok I see it now.